crates/red_knot_python_semantic/resources/mdtest/call/methods.md

```diff
@@ -0,0 +1,192 @@
+        return "a"
+```
+
+Whenever we access the `f` attribute through the class object itself (`C.f`) or through an instance
+(`C().f`), this access happens via the descriptor protocol. Functions are (non-data) descriptors
+because they implement a `__get__` method. This is crucial in making sure that method calls work as
+expected. In general, the signature of the `__get__` method in the descriptor protocol is
+`__get__(self, instance, owner)`. The `self` argument is the descriptor object itself (`f`). The
+passed value for the `instance` argument depends on whether the attribute is accessed from the class
+object (in which case it is `None`), or from an instance (in which case it is the instance of type
+`C`). The `owner` argument is the class itself (`C` of type `Literal[C]`). To summarize:
+
+- `C.f` is equivalent to `getattr_static(C, "f").__get__(None, C)`
+- `C().f` is equivalent to `getattr_static(C, "f").__get__(C(), C)`
+
+Here, `inspect.getattr_static` is used to bypass the descriptor protocol and directly access the
+function attribute. The way the special `__get__` method *on functions* works like is as follows. In
+the former case, if the `instance` attribute is `None`, it simply returns the function itself. In
+the latter case, it returns a *bound method* object:
+
+```py
+from inspect import getattr_static
+
+reveal_type(getattr_static(C, "f"))  # revealed: Literal[f]
+
+reveal_type(getattr_static(C, "f").__get__)  # revealed: <method-wrapper `__get__` of `f`>
+
+reveal_type(getattr_static(C, "f").__get__(None, C))  # revealed: Literal[f]
+reveal_type(getattr_static(C, "f").__get__(C(), C))  # revealed: <bound method `f` of `C`>
+```
+
+In conclusion, this is why we see the following two types when accessing the `f` attribute on the
+class object `C` and on an instance `C()`:
+
+```py
+reveal_type(C.f)  # revealed: Literal[f]
+reveal_type(C().f)  # revealed: <bound method `f` of `C`>
+```
+
+A bound method is a callable object that contains a reference to the `instance` that it was called
+on (can be inspected via `__self__`), and the function object that it refers to (can be inspected
+via `__func__`):
+
+```py
+bound_method = C().f
+
+reveal_type(bound_method.__self__)  # revealed: C
+reveal_type(bound_method.__func__)  # revealed: Literal[f]
+```
+
+When we call the bound method, the `instance` is implicitly passed as the first argument (`self`):
+
+```py
+reveal_type(C().f(1))  # revealed: str
+reveal_type(bound_method(1))  # revealed: str
+```
+
+When we call the function object itself, we need to pass the `instance` explicitly:
+
+```py
+C.f(1)  # error: [missing-argument]
+
+reveal_type(C.f(C(), 1))  # revealed: str
+```
+
+When we access methods from derived classes, they will be bound to instances of the derived class:
+
+```py
+class D(C):
+    pass
+
+reveal_type(D().f)  # revealed: <bound method `f` of `D`>
+```
+
+If we access an attribute on a bound method object itself, it will defer to `types.MethodType`:
```

Comment:
Fun Python esoteric knowledge that is probably irrelevant here, but which you may find interesting: if there are extra attributes present in a function's instance `__dict__` (e.g. attributes that have been monkey-patched onto the function), those attributes are also accessible from a bound method derived from the function. E.g.

```pycon
>>> def foo(self): pass
... 
>>> foo.bar = 42
>>> method = foo.__get__(object())
>>> method
<bound method foo of <object object at 0x1022307b0>>
>>> method.bar
42
>>> foo.bar = 56
>>> method.bar
56
```

I'm not sure exactly whether attributes are first looked up in the underlying function's instance dictionary or the class dictionary of `types.MethodType`.

Anyway, probably irrelevant for us here, as I say!

Reply:
Yes. Bound methods simply keep a reference to the function in a `im_func` member …

https://github.com/python/cpython/blob/10b32054ad6bce821e3b40101d4414d025be6e36/Objects/classobject.c#L124

… which they expose via `bound_method.__func__` here …

https://github.com/python/cpython/blob/10b32054ad6bce821e3b40101d4414d025be6e36/Objects/classobject.c#L161C32-L161C38

… and they fall back to looking up attributes on the function in `getattr` here:

https://github.com/python/cpython/blob/10b32054ad6bce821e3b40101d4414d025be6e36/Objects/classobject.c#L212

Reply:
Ah yes, and of course that's visible from Python if you do something like this -- the attribute in the `MethodType` class dictionary takes precedence over the attribute in the function's instance dictionary!

```pycon
>>> def foo(): ...
... 
>>> foo.__hash__ = 42
>>> method = foo.__get__(object())
>>> method.__hash__
<method-wrapper '__hash__' of method object at 0x102ed0a40>
```

I was obviously lacking imagination last night 😄